libbeat/testing/integration/output_watcher.go

```diff
@@ -185,3 +189,58 @@ func (w *metaWatcher) String() string {
 	}
 	return " * " + strings.Join(expectations, "\n * ")
 }
+
+// NewJSONWatcher watches for existence of certain key-value pair in a JSON document
+func NewJSONWatcher(fields mapstr.M) OutputWatcher {
+	return &jsonWatcher{
+		fields: fields,
+	}
+}
+
+type jsonWatcher struct {
+	fields mapstr.M
+}
+
+func (w *jsonWatcher) Inspect(line string) {
+	if w.Observed() {
+		return
+	}
+
+	var outputDoc mapstr.M
+	err := json.Unmarshal([]byte(line), &outputDoc)
+	// If a line is not json type, we ignore the error and move ahead
+	if err != nil {
+		return
+	}
+
+	outputDoc = outputDoc.Flatten()
+
+	for key, value := range w.fields {
+		if v, ok := outputDoc[key]; ok {
+			switch v.(type) {
+			case int, string, float64:
+				if v != value {
+					return
+				}
+			default:
+				if reflect.DeepEqual(v, value) {
+					return
+				}
+
+			}
+		}
+
+		w.fields = nil
+	}
```

Comment:
I like the idea, however, I think we could make it simpler to use. If use our [custom `mapstr` implementation](https://github.com/elastic/elastic-agent-libs/blob/main/mapstr/mapstr.go) and support dot-separated deep paths without having to define nested maps in tests.

If we do this, we should update the docs on the `ExpectJSONFields` function as well.